crates/compose/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+    // manifest.
+    manifest_name: DependencyName,
+    // The instantiation id for the dependency node.
+    instantiation_id: NodeId,
+    // The world id for the dependency node.
+    world_id: WorldId,
+    // Name of optional export to use to satisfy the dependency.
+    export_name: Option<String>,
+}
+
+fn apply_deny_all_adapter(
+    dependency_name: &str,
+    dependency_source: &[u8],
+) -> anyhow::Result<Vec<u8>> {
+    const SPIN_VIRT_DENY_ALL_ADAPTER_BYTES: &[u8] = include_bytes!("../deny_all.wasm");
+    let mut graph = CompositionGraph::new();
+
+    let dependency_package =
+        Package::from_bytes(dependency_name, None, dependency_source, graph.types_mut())?;
+
+    let dependency_id = graph.register_package(dependency_package)?;
+
+    let deny_adapter_package = Package::from_bytes(
+        "spin-virt-deny-all-adapter",
+        None,
+        SPIN_VIRT_DENY_ALL_ADAPTER_BYTES,
+        graph.types_mut(),
+    )?;
+
+    let deny_adapter_id = graph.register_package(deny_adapter_package)?;
+
+    match wac_graph::plug(&mut graph, vec![deny_adapter_id], dependency_id) {
+        Err(wac_graph::PlugError::NoPlugHappened) => {
+            // Dependencies may not depend on any interfaces that the plug fills so we shouldn't error here.
+            // Just return the origin `dependency_source` as is.
+            return Ok(dependency_source.to_vec());
+        }
+        Err(other) => {
+            anyhow::bail!(
+                "failed to plug deny-all adapter into dependency: {:?}",
+                other
+            );
+        }
+        Ok(_) => {}
+    }
+
+    let bytes = graph.encode(Default::default())?;
+    Ok(bytes)
+}
+
+/// Returns true if the dependency name matches the provided import name string.
+fn matches_import(dependency_name: &DependencyName, import_name: &str) -> bool {
+    // If the import name does not start with the package name, it cannot match.
+    if !import_name.starts_with(&dependency_name.package.to_string()) {
+        return false;
+    }
+    // Check for exact package/interface/version match.
+    if import_name == dependency_name.to_string() {
+        return true;
+    }
+
+    if dependency_name.interface.is_none() {
+        // Package wildcard (i.e. interfaceless).
+        // If the dependency name has a version, it must match the version specified in the import name.
+        if let Some(version) = dependency_name.version.as_ref() {
+            import_name.ends_with(&version.to_string())
+        } else {
+            // Otherwise, the dependency name has no version and matches any version.
+            true
+        }
+    } else {
+        // Interface wildcard (i.e. versionless).
+        if dependency_name.version.is_none() {
+            // We have an interface wildcard (i.e. versionless) where package/interface must match.
+            import_name.starts_with(&dependency_name.to_string())
```

Comment:
I think this isn't quite right. This makes dependency matching work sort of like a wildcard I believe. For example, in the example app, the following works without problem:

```
"component:animal-facts/animal" = { path = "target/wasm32-wasi/release/animal_facts_dep.wasm" }
```

Notice the dependency is `"component:animal-facts/animal"` and not the expected `"component:animal-facts/animal-facts"`

Reply:
Great catch!

Reply:
Updated and simplified, PTAL.